jdaviz/configs/default/plugins/subset_tools/subset_tools.py

```diff
@@ -224,61 +224,41 @@ def get_regions(self, region_type=None, list_of_subset_labels=None,
-                subset_label = subset_data.label
+        # first get ALL subsets of specified spatial/spectral type(s)
+        subsets = self.app.get_subsets(spectral_only=region_type == ['spectral'],
+                                       spatial_only=region_type == ['spatial'],
+                                       include_sky_region='sky' in reg_type,
+                                       use_display_units=use_display_units)
 
-                if isinstance(subset_data.subset_state, MaskSubsetState):
-                    # Ignore MaskSubsetState here
-                    continue
+        labels = list_of_subset_labels or subsets.keys()
+        labels = [labels] if not hasattr(labels, '__iter__') else labels
 
-                try:
-                    if self.app.config == "imviz" and to_sky:
-                        region = roi_subset_state_to_region(subset_data.subset_state,
-                                                            to_sky=to_sky)
-                    else:
-                        region = subset_data.data.get_selection_definition(
-                            subset_id=subset_label, format='astropy-regions')
-                except (NotImplementedError, ValueError):
-                    failed_regs.add(subset_label)
+        regions, failed_regs = {}, set()
+        for subset_label in labels:
+            try:
+                ss = subsets[subset_label]
+                if isinstance(ss, SpectralRegion):
+                    regions[subset_label] = ss
                 else:
-                    regions[subset_label] = region
-
-            if len(failed_regs) > 0:
-                self.app.hub.broadcast(SnackbarMessage(
-                    f"Regions skipped: {', '.join(sorted(failed_regs))}",
-                    color="warning", timeout=8000, sender=self.app))
-
-        if 'spectral' in region_type:
-            spec_regions = self.app.get_subsets(spectral_only=True,
-                                                use_display_units=use_display_units)
-            if spec_regions:
-                regions.update(spec_regions)
-
-        # filter by list_of_subset_labels
-        if list_of_subset_labels is not None:
-            regions = {key: regions[key] for key in list_of_subset_labels}
+                    reg = _chain_regions([x[reg_type] for x in ss],
+                                         [x['glue_state'] for x in ss])
+                    if reg is None:
+                        failed_regs.add(subset_label)
+                    else:
+                        regions[subset_label] = reg
+            except ValueError:  # key doesnt exist, subset wasn't retrieved
+                failed_regs.add(subset_label)
+
+        if len(failed_regs) > 0:
+            self.app.hub.broadcast(SnackbarMessage(
+                                  f"Regions skipped: {', '.join(sorted(failed_regs))}",
+                                  color="warning", timeout=8000, sender=self.app))
```

Comment:
If this is generally going to be called via the API it might be better to raise a warning here, I often find if I'm doing something in code cells any snackbar messages will have timed out by the time I look back at the app.

Reply:
done!